src/pages/Governance/Proposal/hooks/useAccountHasVoted.tsx

```diff
@@ -0,0 +1,71 @@
+import {useEffect, useState} from "react";
+
+import {getTableItem} from "../../../../api";
+import {useGetAccountResource} from "../../../../api/hooks/useGetAccountResource";
+import {useWalletContext} from "../../../../context/wallet/context";
+import {GlobalState, useGlobalState} from "../../../../GlobalState";
+
+const fetchTableItem = async (
+  pool_address: string,
+  proposal_id: string,
+  handle: string,
+  state: GlobalState,
+): Promise<boolean> => {
+  const tableItemRequest = {
+    key_type: "0x1::aptos_governance::RecordKey",
+    value_type: "bool",
+    key: {
+      stake_pool: pool_address,
+      proposal_id,
+    },
+  };
+
+  await getTableItem(
+    {tableHandle: handle, data: tableItemRequest},
+    state.network_value,
+  );
+
+  return true;
+};
+
+interface VotingRecords {
+  votes: {
+    handle: string;
+  };
+}
+
+interface OwnerCapability {
+  pool_address: string;
+}
+
+export function useAccountHasVoted(proposalId: string): boolean {
+  const [state, _setState] = useGlobalState();
+  const {accountAddress} = useWalletContext();
+  const [hasVoted, setHadVoted] = useState<boolean>(false);
+
+  const {accountResource: ownerCapabilityResource} = useGetAccountResource(
+    accountAddress || "0x1",
+    "0x1::stake::OwnerCapability",
+  );
+
+  const {accountResource: votingRecordResource} = useGetAccountResource(
+    "0x1",
+    "0x1::aptos_governance::VotingRecords",
+  );
+
+  useEffect(() => {
+    if (!votingRecordResource || !ownerCapabilityResource) return;
+
+    const votingResource = votingRecordResource.data as VotingRecords;
+    const {handle} = votingResource.votes;
+
+    const ownerCapability = ownerCapabilityResource.data as OwnerCapability;
+    const {pool_address} = ownerCapability;
+
+    fetchTableItem(pool_address, proposalId, handle, state)
+      .then(setHadVoted)
+      .catch(() => setHadVoted(false)); // fetchTableItem returns a 404 error if the item is not in the table
```

Comment:
What if there's another error such as a 500?

Reply:
Good point! in that case it is a server error and we should also set it to false - I think the bigger question is if there is a server error in (any) of the requests, what should we display? I will address it in a later PR